ui-src/lib/models/JsonWrapper.js

```diff
@@ -0,0 +1,51 @@
+// vim: set et sw=2:
+//
+
+/**
+ * WeakMap that keeps our private data.
```

Comment:
Same question about WeakMap here.

Reply:
The reason is that the JSON is now kept outside the class itself, so cannot be discovered by class users. The whole intent of this class is to hide the JSON representation.

When we're all done incorporating this change and things are working, I intend to remove the currently existing json() getter to make it truly inaccessible. We'll, a truly dedicated individual can still find it...

Reply:
As I asked in the other comment, doesn't keeping this in a WeakMap mean it can be thrown away on garbage collection? If not, what keeps the map entries alive? There's no other reference to the values. I think all the models can simply stop working on GC, including subclass methods that do things like `this.json.uuid`. I don't even know how to test this, as Chrome apparently doesn't have a way to force GC. (And if it doesn't throw them away, then it just leaks memory, as nothing explicitly deletes entries from the map.)

I'd much rather just follow the [style guide's rule for private fields](https://google.github.io/styleguide/jsguide.html#features-classes-fields). That's enough to make it clear they shouldn't be used from outside the class, and anything else more strikes me as just obfuscation and potential bugs.

Reply:
I stand corrected on the WeakMap. I thought it was a weak reference to the values, but you're saying it's a weak reference to the keys, and I agree those are referenced elsewhere.

Still, I don't see the point. I looked up the extending an object externally you mentioned. I think externally is a key word; I didn't see anything advocating doing this from the object itself. And I agree that callers could just ignore the _ and use private stuff anyway, but as you pointed out that's also true of the WeakMap. It's impossible to for the callee to prevent callers from abusing it if they're determined enough.

Reply:
ping - as mentioned, I don't still don't think the WeakMap is accomplishing anything

Reply:
It is accomplishing hiding the internal implementation detail to the (casual) observer of the code, or the object in the debugger. This discourages trying to use this detail, more so than obviously visible members with a _ suffix. I think here, and in other comments, I have expressed my point of view fairly clearly. As such, this serves a purpose, and introducing several "private" variables does not do any better.

I suppose all this underlies my past experience with OO programming and data hiding practices.
These articles describes what I did, and why they believe it is valuable:

http://fitzgeraldnick.com/2014/01/13/hiding-implementation-details-with-e6-weakmaps.html
https://www.sitepen.com/blog/2015/03/19/legitimate-memory-efficient-privacy-with-es6-weakmaps/
http://lanceball.com/words/2016/10/14/data-hiding-in-es6/
etc.

Would you feel better if the convenience getter "json" was not there? That's the only point of external visibility.

I think I've spent enough time on this topic now.

Reply:
Thanks for the article links. This level of obfuscation still seems strange to me; and I still think those articles are describing a different situation (third-party extension). But we both really want to get this merged, and I've heard you want to get eventually get rid of the json in this WeakMap anyway.